src/main/perl/Lock.pm

```diff
@@ -73,237 +65,155 @@ The B<CAF::Lock> class provides methods for handling application locking.
+If I<force> is set to B<FORCE_ALWAYS> then the lock file is just set
+again, even if the lock is already set by another application
+instance, and neither I<timeout> nor I<retries> are taken
+into account.
 
 =cut
 
 sub set_lock {
-  my ($self,$retries,$timeout,$force)=@_;
-
-  $retries=0 unless (defined $retries);
-  $timeout=0 unless (defined $retries);
-
-  if ($self->{LOCK_SET}) {
-    # oops.
-    $self->error("lock already set by this application instance");
-    return undef;
-  }
-
-  unless ($force == FORCE_ALWAYS ||
-	  ($force == FORCE_IF_STALE && $self->is_stale())) {
-    my $tries=1;
-    while ($tries <= $retries &&
-	   ($self->is_locked() && !($force == FORCE_IF_STALE && $self->is_stale()))) {
-      my $sleep=rand($timeout);
-      $self->verbose("lockfile is already held, try $tries out of $retries");
-      $tries++;
-      sleep($sleep);
+    my ($self, $retries, $timeout, $force) = @_;
+
+    $retries = 0 unless (defined $retries);
+    $timeout = 0 unless (defined $retries);
+
+    if ($self->{LOCK_SET}) {
+        # oops.
+        $self->error("lock already set by this application instance: $self->{LOCK_FILE}");
+        return;
     }
-  }
-  if ($self->is_locked() && !($force == FORCE_ALWAYS ||
-	  ($force == FORCE_IF_STALE && $self->is_stale()))) {
-    $self->error("cannot acquire lock file: ".$self->{'LOCK_FILE'});
-    return undef;
-  }
-  # now got the lock (or forcing!)
-  my $lf=FileHandle->new("> " . $self->{'LOCK_FILE'});
-  unless ($lf) {
-    $self->error("cannot create lock file: ".$self->{'LOCK_FILE'});
-    return undef;
-  }
-  print $lf $$;
-  unless ($lf->close()) {
-    $self->error("cannot close lock file: ".$self->{'LOCK_FILE'});
-    return undef;
-  }
-  $self->{LOCK_SET}=1;
-  return SUCCESS;
-}
 
+    my $tries = 0;
+    do {
+        if ($tries > 0) {
+            $self->verbose("lock file is already held, try $tries out of $retries");
+            sleep($timeout);
+        }
+        $tries++;
+        return SUCCESS if $self->_try_lock($force);
+    } while ($tries < $retries && $timeout);
+
+    $self->error("cannot acquire lock: $self->{LOCK_FILE}");
+    return;
+}
 
 =pod
 
-=item unlock
+=item unlock()
 
-Releases the lock and returns SUCCESS. Reports an error and returns
-undef if the lock file cannot be released. If the object (application
-instance) does not hold the lockfile, an error is reported and undef
+Releases the lock and returns B<SUCCESS>.  Reports an error and returns
+B<undef> if the lock cannot be released.  If the object (application
+instance) does not hold the lock, an error is reported and B<undef>
 is returned.
 
 =cut
 
 sub unlock {
-  my $self=shift;
-  if ($self->{LOCK_SET}) {
-    unless (unlink($self->{'LOCK_FILE'})) {
-      $self->error("cannot release lock file: ",$self->{'LOCK_FILE'});
-      return undef;
+    my $self = shift;
+    if ($self->{LOCK_SET}) {
+        # if we forced the lock LOCK_FH can be undef
+        if ($self->{LOCK_FH}) {
```

Comment:
if `$self->{LOCK_FH}` can be 0 or `''`, this is not going to work. better use `if (defined($self->{LOCK_FH}))`